Many GLSL and MSL fixes and improvements

zeux opened the pull request:
Let me preface this by saying that since Thekla/hlslparser is sort of Witness specific, feel free to reject my pull request if you think you guys don't need these changes :) I also apologize for the size of the request - this is an accumulation of many commits in my repository but since I had to do pretty crazy merges from Thekla with whitespace changes etc. I don't know if I can split it into small chunks easily. Finally, while this is mostly strict improvements, it does change the output shaders a bit wrt the contract between the API and shaders - I summarized what I think are breaking changes at the end.

Here's the summary of all changes this contains:

General code improvements:
- Various warning fixes
- Split WriteLineTagged from WriteLine for file/line overloads - existing overload makes it impossible to use "%s %d" format since the file/line overload is mistakenly selected
- More cleanly add printf attributes
- Add Travis CI config and a Makefile
- Expand command-line tool to support more output languages
- Fix VS2012 compatibility - do not rely on field initializers inside struct definitions

GLSL improvements/fixes:
- Fix #line directives to avoid the file name which does not conform to the standard
- Add support for various GLSL versions (GLSL 2 & 3, GLSL ES 2 & 3). This required tweaks throughout the codegen since the resulting syntax is frequently somewhat different.
- Add texture array support
- Use more precise filtering for the output functions - instead of checking if string pool contains a name check if there's a function that is used
- Implement implicit casting (HLSL-like) in ternary ifs and some other operators
- Implement support for MRT color output and generally improve built-in semantic support (add D3D9 semantics like POSITION, add PSIZE, VPOS, etc.)
- Improve attribute/varying generation - code now works with output arguments (out float4 pos: POSITION), better struct support, etc.
- Move gl_Position flipping behind a separate codegen flag (FlipPosition)
- Add a flag to switch between column-major and row-major matrices - currently only affects matrix multiplication...
- Add a flag to emulate constant buffers - this generates a flat uniform array vec4 ...[N] and generates optimal statically or dynamically computed indices into that. This helps with compiling for GL 2/ GLES 2.
- Add a flag to lower matrix multiplication to dot products or mul/mad - this improves codegen after glsl-optimizer because it's able to better simplify certain expressions
- Add tex2Dgrad support
- Add ddx/ddy support
- Add function forward declaration support

MSL improvements/fixes:
- General translation for constant buffers instead of hardcoding per_item/etc.
- Add support for explicit sampler and constant buffer register specification (: register(s1), etc.)
- Some fixes for output filtering based on whether we think we need certain types/functions - rely on NeedsFunction and fix some corner cases
- Add function forward declaration support
- Add float4x4-to-float3x3 cast support
- Add all pairs of square-matrix-by-vector multiplication

Likely regressions on existing shaders:
- Took out float3x4/float4x3 support from MSL and GLSL. This is probably a mistake and I can restore it.
- Restructured hard-coded MSL vertex attribute mapping. This is probably something that needs support for external configuration
- Default GLSL flags disable gl_Position flipping. I believe this is the right call and caller should specify flipping if it is needed
- MSL shaders may get bigger as far as output source code is concerned unless the tree is pruned using PruneTree. I believe this is the right call.
- Removed EmulateAlphaTest. This is probably a mistake - I forgot the reason why I did this :) I can restore it.

zeux committed:
Fix Mac build

zeux committed:
Add support for ddx/ddy

zeux committed:
Add support for tex2Dgrad

zeux committed:
Add support for tex3D

zeux committed:
GLSL spec says #line can't have a file name so don't output it

zeux committed:
Insert casts into expressions like (a < 0) * 2.0

Multiplication/division did not propagate destination types before

zeux committed:
Fix tex2Dbias compilation

zeux committed:
Fix generation of pruned trees

We no longer output hidden declarations

zeux committed:
Add support for 'out' attributes in entry point

Now we correctly declare output attributes if they are specified in the entry
point, pass them to the function and handle void return type from the shader
properly.

zeux committed:
Fix Linux build

zeux committed:
Add Makefile and Travis config

zeux committed:
Travis CI should really switch to current Ubuntu

zeux committed:
Add HLSL generation to command line tool

New command line arguments control the target language:

-glsl - GLSL
-hlsl - DX10-style HLSL
-legacyhlsl - DX9-style HLSL

zeux committed:
Treat built-in semantics differently between input/output

The same identifier can mean different things, e.g. SV_POSITION output from VS
is gl_Position but SV_POSITION input in FS is gl_FragCoord.

zeux committed:
Treat POSITION as a valid VS output

zeux committed:
Implement support for zero-initializing structs

A common idiom in HLSL is being able to cast 0 to a struct to zero-initialize.
GLSL does not support this so we need to expand the cast to explicitly enumerate
all arguments.

zeux committed:
Implement support for sampler2DArray and tex2DArray

zeux committed:
Implement forward declaration support

It's not too pretty unfortunately - there are now two function instances, and
one forwards to another. Not sure if there's a better way to do this - we can
split HLSLFunction declaration and definition nodes but we need to link
FunctionCall instances to something, and to keep forward declaration and the
definition nodes visible if we want to translate all of them.

Alternatively we could try to just automatically insert forward declaration
nodes in the output...

zeux committed:
Fix output argument processing

Directly outputting arguments by passing them to out args of entrypoint does
not work if we need to post-process the result - OutputSetOutAttribute has
to be called.

zeux committed:
Implement support for mul(matrix, vector)

zeux committed:
Introduce preliminary support for other GL versions

This adds the infrastructure for version selection, correct version declaration
and also support for correct attribute qualifiers for old GL versions.

zeux committed:
Simplify tex2Darray implementation

zeux committed:
Simplify GL version logic

Split versions into "legacy" and not

zeux committed:
Support texture functions for legacy profiles

zeux committed:
Fix uint/uvecN types on legacy desktop GL

zeux committed:
Stop emitting functions we don't need for pruned trees

GetContainsString is too crude - it does a string pool lookup. Instead check
if the function we need is actually used.

Since this is mostly used for intrinsics, we can't just call FindFunction and
have to do the tree traversal instead.

zeux committed:
Add printf format specifiers to CodeWriter

zeux committed:
Fix some compilation warnings

zeux committed:
Implement scalar type upcast for conditionals

zeux committed:
Disable position Y flipping by default

Allow to override this with GLSL generation flags

zeux committed:
Add compatibility shims for texture functions on old GLSL versions

Also enable a GLES extension if dFdx/dFdy are used.

zeux committed:
Add mat4 -> mat3 cast hack

Hopefully we can actually set the baseline to OpenGL 2.1 and not worry about
this but for now let's just do it.

zeux committed:
Merge remote-tracking branch 'upstream/master'

Whitespace changes were ignored during the merge

zeux committed:
Convert all remaining whitespaces to LF

The previous whitespace merge destroyed the history anyway!

zeux committed:
Fix MSLGenerator build

zeux committed:
Merge branch 'master' of https://github.com/zeux/hlslparser

zeux committed:
Fix MSLGenerator compilation for VS2012

zeux committed:
Fix Linux build

zeux committed:
Merge branch 'master' of github.com:zeux/hlslparser

zeux committed:
GLSL: Improve fragment shader output translation

Unfortunately, OpenGL makes it really hard to use MRT in certain versions.
Specifically, you can't output location attributes before GL3.3.

Thus we have to declare an appropriately sized array manually...

zeux committed:
Add Metal support to the command-line compiler

zeux committed:
GLSL: Implement constant buffer emulation

Emit a fake struct for the constant buffer contents and packing code that follows
HLSL layout rules.

This means we can setup constants using glUniformfv even if the target does not
support uniform buffers.

zeux committed:
GLSL: Fix cbuffer translation for array access

It's correct to route inside the identifier access (obviously)

zeux committed:
GLSL: Add constant buffer emulation for vec4 arrays

Normally we don't support arrays because array literals are not supported by
legacy GLSL, but if there's just one vec4 array it's easy to just return it.

At some point we might implement compile-time evaluation of field indexing or
array access, but it's much easier to use a struct getter for now.

zeux committed:
GLSL: Implement row-major matrix packing

Currently this only changes the multiplication order, whereas in reality
this should probably affect cbuffer unpacking...

zeux committed:
GLSL: Minor cleanup

zeux committed:
GLSL: Implement matrix multiplication lowering

Matrix operation are lowered into a sequence of dot products or mul/add
operations. This helps glsl-optimizer do further optimizations because it
does not do this lowering by itself.

zeux committed:
Fix WriteLine overload sadness

You can't use a format string "%s %d" right now without compiler switching to
file/line overload...

zeux committed:
GLSL: Switch to manual cbuffer layout

Instead of relying on GLSL struct packing rules, manually create the necessary
expressions to index the resulting array.

The expressions are somewhat tricky to construct because they are split into
static and dynamic indexing, and static indexing requires sub-indexing the
resulting vec4 for small types.

zeux committed:
GLSL: Optimize static array indexing

Perform early constant folding for array indexing if possible

zeux committed:
MSL: Implement cast-0-to-struct

Also restrict the scope of GLSL zero-literal output to user-defined types.

zeux committed:
MSL: Output correct uniform buffer declarations

Buffer offset still looks like a gross hack but the buffers are otherwise
properly bound regardless of their name.

zeux committed:
MSL: Implement better texture/buffer register assignment

We now parse the source register index and use it for buffers and textures.
If no index is specified, we allocate the next available register index.

This is no longer doing an offset for constant buffers, assuming that C++ code
can use a fixed offset for all vertex buffers...

zeux committed:
MSL: Add more mul overloads

zeux committed:
MSL: Fix texturing functions

Also fix forward declarations - no need to emit them because class members are
visible to each other regardless of the order.

zeux committed:
MSL: Remove redundant includes

zeux committed:
GLSL: Remove zero-literal casting support

It's cleaner to not use it in the shaders anyway, and it requires special
casing for both GLSL and MSL... Just disable it.

zeux committed:
MSL: Update semantic translation function

Support D3D10 semantics, remap attribute inputs in a more uniform way.

zeux committed:
MSL: Various generation improvements

Stop using FindFunctionCall - it does not work properly because it does not
guard function definition emission. The correct approach is to use
NeedsFunctionCall, and to pre-prune a tree if necessary.

Always emit texture structures - they don't hurt anything, and without
these it's impossible to have a function that accepts a sampler and does not
use it...

Also add float4x4 -> float3x3 casting support.

zeux committed:
MSL: Add support for row-major matrix packing

Also re-introduce the vertex buffer offset for now, it's not super clear whether
we can avoid it.

zeux committed:
GLSL: Implement point size support

zeux committed:
MSL: Add point_size support

zeux committed:
Kill stray __attribute__ define

zeux committed:
Whitespace cleanup

This reduces the number of diffs against upstream

zeux committed:
More whitespace cleanup

zeux commented:
Also if you have no interest in these changes, please let me know - I'll detach my fork from this repository and maintain it separately.

castano commented on src/MSLGenerator.cpp:
```diff
@@ -210,61 +210,54 @@ void MSLGenerator::Prepass(HLSLTree* tree, Target target, HLSLFunction* entryFun
-    // @@ IC: Hack! LIST BUFFERS IN THE RIGHT order.
-    int bufferOffset = 0;
-    if (target == Target_VertexShader) {
-        bufferOffset = 1; // Index 0 reserved for input vertex buffer.
```
When using incremental register assignment, it would still be good to reserve 0 for vertex buffer, or better, to provide a way for the application to reserve buffer indices.

zeux replied:
Added MSLGenerator::Options::bufferRegisterOffset to control this.

castano replied:
Neat!

castano commented:
I'd be very interested in merging your changes and continue to do so at more regular intervals in the future. However, as you point out we would need to fix the regressions for us to be able to share the code.

Some of these are witness-specific (like the use of two uniform buffers or the buffer register assignment). I would like to add them back, not necessarily by default, but through some config options or flags. Others are more general, like the support for 4x3 matrices.

I'll review your changes in more detail and add some comments.

castano commented on src/MSLGenerator.h:
```diff
@@ -88,10 +88,6 @@ class MSLGenerator
     bool            m_error;
-    
-    // TODO: Temporary hack to access uniforms struct members
-    HLSLBuffer* m_per_pass_buffer;
```
Splitting uniform parameters in two buffers based on the frequency of change is critical for performance. I would imagine not every game cares about this, or would want to do it the same way, but I would certainly like to keep this feature, at least as an optional code path.

zeux replied:
Sure - but there is no need to special-case these two names in the translator. My understanding is that you already use something like `cbuffer per_item { float4 field1; ... }` - with this change the only other change you'll need to keep bindings the same is specify a register, like `cbuffer per_item: register(b0) { float4 field1; }` (otherwise it's first come first served, as it is with textures in your version). If you parse the argument table on the Metal side then it'll just work without any changes I think - per_item/per_pass would be assigned to different slots (see ParseRegister)

zeux replied:
The reason for this change is that we obviously don't use per_item/per_pass names :) And we use up to 4 constant buffers in some shaders - so I made the code generic. I believe in this case it actually preserves behavior - I could run some shaders that do use per_item/per_pass cbuffer names and confirm this if you want.

castano replied:
We have a lot of D3D9 shaders and changing their code is not easy, since they are used in other platforms. The way we segregate the uniforms is by tagging them with a special semantic. I'll see if there's a cleaner and more general way of supporting this, but I'll certainly have to bring this feature back in one way or another.

castano replied:
What about a 'buffer(name)' semantic for uniforms not explicitly declared inside a buffer? That way the uniform is added to the named buffer or it's automatically created if it does not exist.

zeux replied:
Ah, I must have missed this - I thought it still used cbuffer declared in shader source, just with a hard-coded name. We use manual cbuffers on all platforms at this point (including D3D9 where they are just ignored by the compiler, and GLES 2 which is what cbuffer emulation in this commit is for). If it's hard to change the shaders to add cbuffers on certain platforms, maybe this should be an optional extension where you just specify the buffer name in some form and the translator automatically creates the necessary cbuffers?

castano commented on src/MSLGenerator.cpp:
```diff
@@ -116,14 +116,17 @@ static const char * translate_input_semantic(MSLGenerator::Target target, const
         if (target == MSLGenerator::Target_VertexShader)
         {
             if (is_semantic(semantic, "POSITION", 0)) return "attribute(0)";
```
I imagine every application may want to do this in a different way. Maybe we can accommodate that with a callback or specifying the mapping up front and storing it in a table?

zeux replied:
Yeah - I'll split it up in a callback or a table.

zeux replied:
Done (with a callback)

castano replied:
Looks good.

castano commented on src/HLSLTree.cpp:
```diff
@@ -1409,139 +1409,5 @@ void HideUnusedArguments(HLSLFunction * function)
 }
 
-
-class FindFunctionCallVisitor : public HLSLTreeVisitor
```
Can you explain why this approach isn't valid? I'm not advocating it, just want to understand. Ideally I think the necessary functions should be listed during traversal and prepended to the generated code, instead of traversing the tree multiple times for each function.

zeux replied:
So as far as I understand, the author of the original code base intended that the translator would translate the AST you have to the target code without really doing dependency analysis. Now, in some cases you want the output shader to be small, or may have some other issues with the approach of translating 1-1 - there's already a function in this code base, PruneTree, that deals with this by marking unreachable elements as hidden.

I have written support for correctly checking if the function is needed when I was improving GLSL support (NeedsFunction), and basically with that FindFunctionCall becomes redundant. They are pretty much the same except that NeedsFunction supports pruned trees by checking for hidden flag and has an early-out by checking for the string in the pool.

I agree that the traversal that is performed multiple times here is inefficient. In context of Metal I've never found it to be a problem - the time clang takes to compile & codegen dwarfs the time hlslparser translation takes by at least an order of magnitude. I can check how much this specific part contributes to the overall runtime.

zeux replied:
P.S. In the original codebase FindFunctionCall performs reachability traversal from the entry point. This may become impossible if we ever add support for multiple entrypoints in one shader source, and this becomes unnecessary if you really don't need the code - you can easily call PruneTree and pass it the entrypoint. I must confess I haven't tested the MSL codegen changes in our compilation stack without pruning - I'll do that to see if there are any correctness issue with the non-pruned tree generation.

zeux replied:
I've checked and our shaders compile fine both with and without pruning (we frequently include redundant definitions and also have some shaders with many many entrypoints, so the resulting .metal files compile much slower, but they do compile). Additionally, I remembered the exact problem which led me to reworking this - it's in the commit description for the specific change that replaces FindFunctionCall with NeedsFunction :) FindFunctionCall was only used as a check "does the shader use tex2D/etc.", but it's possible that there's a function that uses tex2D but is not called by the shader entrypoint - without the tree pruning this function would be emitted and fail to compile since tex2D wouldn't be.

castano replied:
Ah, I see. That makes sense!

castano commented on src/MSLGenerator.cpp:
```diff
@@ -222,6 +222,9 @@ void MSLGenerator::Prepass(HLSLTree* tree, Target target, HLSLFunction* entryFun
     HLSLType samplerType(HLSLBaseType_Sampler);
 
+    // Reserve buffer 0 for implicit vertex buffer
+    int bufferRegisterOffset = (m_target == Target_VertexShader) ? 1 : 0;
```
Ah, glad you added that back. Ideally this should be generator option. Some use cases may have multiple vertex buffers, others may not have any, etc.

zeux replied:
Yes, I'm not very happy about this myself. Maybe it should be an offset you can pass to the generator and you'd be supposed to pass in 1 manually - I was also thinking for our code of removing the offset and trying to use buffers at a fixed offset for vertex streams (e.g. slots 8+ would be for vertex buffers...).

zeux replied:
Added MSLGenerator::Options::bufferRegisterOffset to control this.

castano commented on src/MSLGenerator.cpp:
```diff
@@ -482,33 +485,32 @@ bool MSLGenerator::Generate(HLSLTree* tree, Target target, const char* entryName
 
     if (m_tree->NeedsFunction("mul")) {
+        const char* am = (flags & Flag_PackMatrixRowMajor) ? "m * a" : "a * m";
+        const char* ma = (flags & Flag_PackMatrixRowMajor) ? "a * m" : "m * a";
```
The implications of row/column major order are much greater than simply the order of the operands in mul. The implementation of operator [] would also have to change, as well as the _ij member accessors. _But_ this is a good start!

zeux replied:
Yeah :( This should be reworked ideally. I think the semantics that packing usually has is that it only affects constant inputs (I'll have to confirm this...) so maybe it's better to transpose the matrix when we read it from the buffers. Unfortunately, for float3x4 it also changes the type which makes it... painful. I'll look into this a bit.

castano commented on src/HLSLTree.cpp:
```diff
@@ -1366,133 +1409,5 @@ void HideUnusedArguments(HLSLFunction * function)
-}
-
-
-bool EmulateAlphaTest(HLSLTree* tree, const char* entryName, float alphaRef/*=0.5*/)
```
Please, add this back!

zeux replied:
Will do! :)

zeux replied:
Added this back, I think I removed this along with FindFunctionCall by mistake.

castano replied:
Thanks!

castano commented on src/HLSLTree.h:
```diff
@@ -328,9 +329,15 @@ struct HLSLStatement : public HLSLNode
 struct HLSLAttribute : public HLSLNode
 {
     static const HLSLNodeType s_type = HLSLNodeType_Attribute;
-    HLSLAttributeType   attributeType = HLSLAttributeType_Unknown;
```
Do you have anything against using C++11 member initialization? Do you have to support old compilers?

zeux replied:
Yup - this is to support VS 2012 which does not support this specific feature.

castano replied:
Ah, I see. It makes life easier, but it's no big deal.

castano commented on src/HLSLTree.cpp:
```diff
@@ -382,6 +382,40 @@ bool HLSLTree::GetExpressionValue(HLSLExpression * expression, int & value)
+    if (!GetContainsString(name))
+        return false;
+
+    struct NeedsFunctionVisitor: HLSLTreeVisitor
```
Consider adding early-outs to terminate traversal as soon as function found.

zeux replied:
My understanding is that the visitor infrastructure doesn't make it too easy. I'll check how much time NeedsFunction takes - it may be worthwhile to skip branches in smth like VisitTopLevelStatement if we already have found our function, although of course for failed lookups - which are probably more common! - it's not beneficial to early out.

zeux replied:
NeedsFunction is completely invisible on the profile (in fact, MSL generation is ~10x faster than HLSL parsing on our shaders - and as mentioned before the actual Metal/clang parsing & generation is much slower than that).

castano replied:
I agree, it's not a perf issue for us either. You are right it doesn't help the most common case. Terminating early is easy, though. Just override the statement traversal.

zeux committed:
MSL: Introduce Options struct

Move flags into options and expose buffer register offset.

The offset is important because in Metal, vertex shader streams are just
buffers and share the namespace with constant buffers etc. Depending on
the application needs it may need to use some slots in the vertex shader
for vertex streams.

zeux committed:
GLSL: Introduce Options struct

Move flags into Options and introduce constant buffer prefix.

By default we use the name from the HLSL source for the constant buffer name in
both emulated and non-emulated paths. In some cases it's important to customize
the buffer name - for example, on some Adreno drivers on Android GLSL linker
can't correctly merge the array sizes and the only way to avoid these bugs is
to have differently named uniform arrays for VS & FS. This option allows for
precisely that, and possibly other use cases.

zeux committed:
Restore EmulateAlphaTest

zeux committed:
Remove unused variables

zeux committed:
MSL: Implement custom attribute callback

Since semantic -> attribute index mapping is very arbitrary, allow the caller
to provide a callback that converts semantic name/index to an attribute index.

Also refactor output/input semantic conversion for other cases, and support
arbitrary number of color outputs instead of just 2.

zeux committed:
Remove unused allocators

zeux committed:
MSL: Move ParseRegister to a free function

It does not need any generator state.

zeux committed:
Restore float3x4 support

zeux commented:
Ok, this is ready for another round of review - I believe I addressed all comments that required code changes, + there's some info in responses to your comments about other issues. Please let me know if I missed anything!

There is still an outstanding problem with the row/column major packing only affecting matrix-vector multiplication - this requires further research as I'm not sure what the optimal implementation strategy is, and I'd prefer to not gate this PR on that.

castano commented:
Looks good! I'll merge the changes, test it in our code base and merge back if there are any more changes necessary on our side.
